File: stdlib/src/main/java/postvm/stdlib/classes/FileClass.java
package postvm.stdlib.classes;

import postvm.Caller;
import postvm.Utils;
import postvm.exceptions.LCLangIOException;
import postvm.library.classes.*;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class FileClass extends LibraryClass {
    public static final PostVMClassPrototype PROTOTYPE = new PostVMClassPrototype(
            "File",
            ObjectClass.PROTOTYPE,
            Utils.listOf(StringClass.PROTOTYPE.type)
    ) {
        @Override
        public int createClass(Caller caller, int[] args) {
            return new FileClass(caller, new File(PostVMClass.instances.get(args[0]).cast(StringClass.class).string)).classId;
        }
    };

    public File file;

    public FileClass(Caller caller, File file) {
        super(caller, PROTOTYPE);
        this.file = file;
    }

    @Override
    public Integer loadGlobal(PostVMClass clazz, String target) {
        switch (target) {
            case "name": return StringClass.get(file.getName());
            case "path": return StringClass.get(file.getPath());
            case "absolutePath": return StringClass.get(file.getAbsolutePath());
            case "getCanonicalPath": return method(StringClass.type, (caller, args) -> {
                try {
                    return StringClass.get(file.getCanonicalPath());
                } catch (IOException e) {
                    throw new LCLangIOException(e.getMessage(), caller);
                }
            });

            case "exists": return method(BoolClass.type, (caller, args) -> BoolClass.get(file.exists()));
            case "isDirectory": return method(BoolClass.type, (caller, args) -> BoolClass.get(file.isDirectory()));
            case "isFile": return method(BoolClass.type, (caller, args) -> BoolClass.get(file.isFile()));
            case "files": return method(ArrayClass.type, (caller, args) -> {
                List<Integer> classList = new ArrayList<>();
                File[] files = file.listFiles();
                if(files!=null) for (File child: files)
                    classList.add(new FileClass(caller, child).classId);

                return new ArrayClass(classList).classId;
            });

            case "openInput": return method(InputClass.PROTOTYPE.type.nullable(), (caller, lcClasses) -> {
                try {
                    return new InputClass(caller, new FileInputStream(file)).classId;
                } catch (FileNotFoundException e) {
                    return NullClass.INSTANCE.classId;
                }
            });

            case "openOutput": method(OutputClass.PROTOTYPE.type.nullable(), (caller, lcClasses) -> {
                try {
                    return new OutputClass(caller, new FileOutputStream(file)).classId;
                } catch (FileNotFoundException e) {
                    return NullClass.INSTANCE.classId;
                }
            });

            case "createDir": return method(BoolClass.type, (caller, lcClasses) -> BoolClass.get(file.mkdir()));
            case "createDirs": return method(BoolClass.type, (caller, lcClasses) -> BoolClass.get(file.mkdirs()));
            case "create": return method(BoolClass.type, (caller, lcClasses) -> {
                try {
                    return BoolClass.get(file.createNewFile());
                } catch (IOException e) {
                    throw new LCLangIOException(e.getMessage(), caller);
                }
            });
        }
        return super.loadGlobal(clazz, target);
    }
}

File: src/main/kotlin/lclang/libs/std/classes/FileClass.kt
package lclang.libs.std.classes

import lclang.LCClass
import lclang.Value
import lclang.constructor
import lclang.lang.ArrayClass
import lclang.lang.StringClass
import lclang.method
import lclang.methods.Method
import lclang.types.NamedType
import lclang.types.Types
import java.io.File

const val FILE_CLASSNAME = "File"
class FileClass(): LCClass(FILE_CLASSNAME) {
    override var constructor: Method = constructor(listOf(Types.STRING), classType) {
        FileClass(File(it[0].toString()))
    }

    constructor(file: File) : this() {
        globals["name"] = StringClass(file.name).asValue()
        globals["path"] = StringClass(file.path).asValue()
        globals["absolutePath"] = StringClass(file.absolutePath).asValue()
        globals["canonicalPath"] = StringClass(file.canonicalPath).asValue()
        globals["exists"] = Value(Types.BOOL) { file.exists() }
        globals["isDirectory"] = Value(Types.BOOL) { file.isDirectory }
        globals["isFile"] = Value(Types.BOOL) { file.isFile }
        globals["files"] = Value(Types.ARRAY) {
            ArrayClass().apply {
                file.listFiles()?.forEach {
                    add(FileClass(it).asValue())
                }
            }
        }

        globals["openInput"] = method (returnType = NamedType(INPUT_CLASSNAME)) { file.inputStream() }
        globals["openOutput"] = method (returnType = NamedType(OUTPUT_CLASSNAME)) { file.outputStream() }

        globals["createDir"] = method (returnType = Types.BOOL) { file.mkdir() }
        globals["createDirs"] = method (returnType = Types.BOOL) { file.mkdirs() }
        globals["create"] = method (returnType = Types.BOOL) { file.createNewFile() }
    }
}